numpy/lib/format.py

```diff
@@ -431,6 +433,19 @@ def _write_array_header(fp, d, version=None):
         header.append("'%s': %s, " % (key, repr(value)))
     header.append("}")
     header = "".join(header)
+    
+    # Check if d is actual array header data or just some random dict as e.g.
+    # in test_large_header. In the first case, add spare space for growth
```

Comment:
I'd rather fix `test_large_header()` instead.

Reply:
Yes, I thought this might be an option too, this is why I've mentioned it in the comment. I'll have a look into it (next push won't fix it, just minor cleanup).

Reply:
Just that I understand correctly: fixing `test_large_header()` implies that `_write_array_header()` can assume that the array is _correct enough_ which implies it must do the same sanity checks e.g. as `_read_array_header()`, which would then imply that `shape` and `fortran_order` is set (otherwise it would not make sense to modify `test_large_header()` and `_write_array_header()` in the first place).
Enabling sanity checks for `_write_array_header()` would need the tests `test_large_header()`, `test_bad_header()` and `test_metadata_dtype()` to be modified as well. While the first one is straight forward, the second one a little ugly and limits the readability of the test case slightly and could make the test case useless in case of future format modifications, the third one is a little challenging because `_write_array_header()` cannot produce certain kinds of invalid arrays anymore.
I've pushed with the `test_metadata_dtype()` failing.
I see three options to resolve the situation:
1. Revert to the old `_write_array_header()` variant that only applies spare space if the header dict is nice enough
2. Fix `test_metadata_dtype()` in a way that assumes that if a test fails, either writing or reading an array can fail as they share a portion of their code. However, if this is not the case in the future anymore or there is an issue in `np.save()` this could render the test useless. 
3. Introduce a context from `contextlib` for `_write_array_header()` to indicate that sanity checks should be skipped. However, I don't know if this is a thing in Numpy and how to properly implement it.
4. Introduce a boolean flag that would allow to write invalid Numpy headers. Probably 3 is less ugly.
What do you think?

Reply:
I think I might just have found a middle ground by skipping the sanity test in `_write_header` and only assuming that `shape` and `fortran_order` are set and modifying tests only so that this works out.

Reply:
I think this could be it, what do you think?